### PR TITLE
Added networks

### PR DESCRIPTION
**Pull request by lewismarshall**

### CODE REVIEW

**Comment by lewismarshall** [comment]
@vaijab @jits @marcinc - all working...
All tested (with flannel default for now.)..:

### kmm output:
```
Apr 24 09:15:45 ip-172-31-10-232.eu-west-2.compute.internal docker[5817]: [apiclient] First node has registered after 4.006554 seconds
Apr 24 09:15:45 ip-172-31-10-232.eu-west-2.compute.internal docker[5817]: [apiconfig] Created RBAC rules
Apr 24 09:15:45 ip-172-31-10-232.eu-west-2.compute.internal docker[5817]: [addons] Created essential addon: kube-proxy
Apr 24 09:15:45 ip-172-31-10-232.eu-west-2.compute.internal docker[5817]: time="2017-04-24T09:15:45Z" level=info msg="Running:kubectl create -f -"
Apr 24 09:15:45 ip-172-31-10-232.eu-west-2.compute.internal docker[5817]: [addons] Created essential addon: kube-dns
Apr 24 09:15:46 ip-172-31-10-232.eu-west-2.compute.internal docker[5817]: time="2017-04-24T09:15:46Z" level=info msg="Releasing lock..."
Apr 24 09:15:46 ip-172-31-10-232.eu-west-2.compute.internal docker[5817]: time="2017-04-24T09:15:46Z" level=info msg="Released lock"
Apr 24 09:15:46 ip-172-31-10-232.eu-west-2.compute.internal systemd[1]: Started Kubernetes Multi-master.
```
### kubectl output:
```
/ # kubectl --namespace=kube-system get pods | grep flannel
kube-flannel-ds-bfwsj                                                 2/2       Running   0          9m
kube-flannel-ds-ztpct                                                 2/2       Running   1          9m
```

File: pkg/Constants/constants.go [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package constants
+
+const (
+	// These can be specified as config to kubeadm and some network providers.
+	DefaultServiceDNSDomain  = "cluster.local"
+	DefaultServicesSubnet    = "10.96.0.0/12"
+	DefaultPodNetwork		 = "10.244.0.0/16"
```
Comment (marcinc) [inline pkg/Constants/constants.go]: indent :)

File: pkg/Constants/constants.go [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package constants
```
Comment (vaijab) [inline pkg/Constants/constants.go]: Package names should be lower case (same goes for directories)

File: pkg/Constants/constants.go [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+
+const (
+	// These can be specified as config to kubeadm and some network providers.
+	DefaultServiceDNSDomain  = "cluster.local"
```
Comment (vaijab) [inline pkg/Constants/constants.go]: These are exported constants, need comments.

File: pkg/kmm/cmd/kmm.go [hunk 4]
```diff
@@ -117,6 +118,7 @@ func getKmmConfig(cmd *cobra.Command) (cfg kmm.Config, err error) {
 		KubeadmCfg: kubeadmConfig,
 		KubePersistentCaCert:	cmd.Flag("kube-ca-cert").Value.String(),
 		KubePersistentCaKey:	cmd.Flag("kube-ca-key").Value.String(),
+		NetworkProvider:		cmd.Flag("network-provider").Value.String(),
```
Comment (vaijab) [inline pkg/kmm/cmd/kmm.go]: You should really get gofmt and the friends set up. :-)

File: pkg/kmm/cmd/network.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+import (
+	"os"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/UKHomeOffice/keto-k8/pkg/kmm"
+	"github.com/spf13/cobra"
+)
+
+// versionCmd represents the version command
```
Comment (vaijab) [inline pkg/kmm/cmd/network.go]: You meant `networkCmd`?

File: pkg/kmm/cmd/network.go [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+	Run: func(c *cobra.Command, args []string) {
+		installNetwork(c)
+	},
+}
+
+func installNetwork(c *cobra.Command) {
+	err := kmm.InstallNetwork(c.Flag("network-provider").Value.String())
+	if err != nil {
+		log.Fatal(err)
+		os.Exit(-1)
```
Comment (vaijab) [inline pkg/kmm/cmd/network.go]: `log.Fatal` calls `os.Exit()` for you.

File: pkg/k8client/client.go [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+)
+
+const CmdKubectl string = "kubectl"
+
+// TODO: Use API (sort out build issues with client lib and kubeadm)
+// Will take a yaml string and deploy it to the API...
+func Create(resource string) (error) {
+	var args = []string {
+		"create",
+		"-f",
```
Comment (jits) [inline pkg/k8client/client.go]: Indentation looks a little awry?

File: pkg/kmm/kmm.go [hunk 8]
```diff
@@ -125,6 +130,14 @@ func CleanUp(cfg Config, releaseLock bool, deleteAssets bool) (err error) {
 	return nil
 }
 
+func InstallNetwork(networkProvider string) (error) {
+	if np, err := network.CreateNetworkProvider(networkProvider); err != nil {
+		return err
```
Comment (jits) [inline pkg/kmm/kmm.go]: If there's an error here, does the network provider creation above need to be rolled back? (or is that just a local and transient thing?)

File: pkg/k8client/client.go (hunk 7 above)
Comment (marcinc) [inline pkg/k8client/client.go]: cosmetic - indentes...

**Comment by lewismarshall** [comment]
@vaijab - OK all in apart from tab - will do a separate PR now for post `gofmt`... I'll squash after you approve...

File: pkg/network/flannel.go [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+            name: kube-flannel-cfg`
+
+	t := template.Must(template.New("flannel").Parse(flannelYaml))
+	var b bytes.Buffer
+	if err := t.Execute(&b, data); err != nil {
+		return b.Bytes(), err
+	}
+
+	return b.Bytes(), nil
+}
```
Comment (marcinc) [inline pkg/network/flannel.go]: NL

File: pkg/network/network.go [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+// Will register providers and set a default provider...
+func init() {
+	Register(NewFlannelNetworkProvider)
+	Register(NewWeaveNetworkProvider)
+}
+
+// Private method for any providers to call without knowledge of k8 client specifics
+func createk8objects(resources string)(error) {
+	return k8client.Create(resources)
+}
```
Comment (marcinc) [inline pkg/network/network.go]: ^^

File: pkg/network/weave.go [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+`
+
+  t := template.Must(template.New("weave").Parse(weaveYaml))
+  var b bytes.Buffer
+  if err := t.Execute(&b, data); err != nil {
+    return b.Bytes(), err
+  }
+
+  return b.Bytes(), nil
+}
```
Comment (marcinc) [inline pkg/network/weave.go]: ^^

**Pull request merged**